sys/kern/kcsan.c

```diff
@@ -0,0 +1,262 @@
+#include <sys/mimiker.h>
+#include <sys/vm.h>
+#include <sys/thread.h>
+#include <machine/interrupt.h>
+#include <sys/klog.h>
+
+#define WATCHPOINT_READ_BIT (1 << 31)
+#define WATCHPOINT_SIZE_SHIFT 28
+#define WATCHPOINT_SIZE_MASK (~(WATCHPOINT_ADDR_MASK | WATCHPOINT_READ_BIT))
+#define WATCHPOINT_ADDR_MASK ((1 << WATCHPOINT_SIZE_SHIFT) - 1)
+
```

Comment:
Could you please add a little bit more comments to non-obvious parts of this file? And if you know a good explanation of the algorithm online (LWN?), a link inside a comment could be helpful for future maintainers of the code.

Reply:
I've added a short description of the algorithm that should clarify some ideas. Please let me know if you find it not explanatory enough.

Reply:
:+1: